src/ForeignKeyEditorDelegate.cpp

```diff
@@ -0,0 +1,162 @@
+            idsComboBox->setCurrentIndex(-1);
+            clauseEdit->clear();
+        });
+
+        connect(tablesComboBox, static_cast<void(QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+                [=](int)
+        {
+            idsComboBox->setCurrentIndex(-1);
+        });
+
+        connect(idsComboBox, static_cast<void(QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+                [=](int index)
+        {
+            bool enableClausesEditor = (index!= -1);
+            clauseEdit->setEnabled(enableClausesEditor);
+        });
+    }
+
+    QString getSql() const
+    {
+        if (tablesComboBox->currentText().isEmpty() || idsComboBox->currentText().isEmpty())
+            return QString();
+        else
+            return QString("%1(%2) %3")
+                .arg(sqlb::escapeIdentifier(tablesComboBox->currentText()))
+                .arg(sqlb::escapeIdentifier(idsComboBox->currentText()))
+                .arg(clauseEdit->text())
+                .trimmed();
+    }
+
+    QComboBox* tablesComboBox;
+    QComboBox* idsComboBox;
+    QLineEdit* clauseEdit; // for ON CASCADE and such
+
+private:
+    QPushButton* m_btnReset;
+};
+
+ForeignKeyEditorDelegate::ForeignKeyEditorDelegate(const DBBrowserDB& db, sqlb::Table& table, QObject* parent)
+    : QStyledItemDelegate(parent)
+    , m_db(db)
+    , m_table(table)
+{
+
+}
+
+QWidget* ForeignKeyEditorDelegate::createEditor(QWidget* parent, const QStyleOptionViewItem& option, const QModelIndex& index) const
+{
+    Q_UNUSED(option)
+    Q_UNUSED(index)
+
+    ForeignKeyEditor* editor = new ForeignKeyEditor(parent);
+    editor->setAutoFillBackground(true);
+
+    const auto objects = m_db.getBrowsableObjects();
+    for (auto obj : objects) {
+        if ("table" == obj.gettype()) {
+            QString tableName = obj.table.name();
+            if (tableName == m_table.name())
+                continue;
+
+            m_tablesIds.insert(tableName, obj.table.fieldNames());
+        }
+    }
+
+    editor->tablesComboBox->addItems(m_tablesIds.keys());
+
+    connect(editor->tablesComboBox, static_cast<void(QComboBox::*)(const QString &)>(&QComboBox::currentIndexChanged),
+            [=](const QString& tableName)
+    {
+        QComboBox* box = editor->idsComboBox;
+        box->clear();
+        box->addItems(m_tablesIds[tableName]);
+        box->setCurrentIndex(-1);
+    });
+
+    return editor;
+}
+
+void ForeignKeyEditorDelegate::setEditorData(QWidget* editor, const QModelIndex& index) const
+{
+    ForeignKeyEditor* fkEditor = static_cast<ForeignKeyEditor*>(editor);
+
+    int column = index.row(); // weird? I know right
+    sqlb::FieldPtr field = m_table.fields().at(column);
+    QSharedPointer<sqlb::ForeignKeyClause> fk = m_table.constraint({field}, sqlb::Constraint::ForeignKeyConstraintType).dynamicCast<sqlb::ForeignKeyClause>();
+    if (!fk.isNull()) {
+        fkEditor->tablesComboBox->setCurrentText(fk->table());
```

Comment:
Hmm sorry, I'm bad at explaining these things... But let me try with an example:
I'm using [this database](https://dev1.dbhub.io/mkleusberg/pr921.db) (you'll need to open it in DB4S, obviously). There is a table aaa and a table bbb. Table bbb has a foreign key to table aaa. Now, when you edit table bbb, you'll see the foreign key constraint like this:
![screenshot_20170109_001146](https://cloud.githubusercontent.com/assets/3153504/21754448/909f9d3c-d600-11e6-909f-1376ed4ea0ad.png)

But when you double click it, you will get this (or at least I do):
![screenshot_20170109_001312](https://cloud.githubusercontent.com/assets/3153504/21754450/998bde56-d600-11e6-86ef-65ac7646c74f.png)
You can see that the columns combo box is empty and that it's grey (=disabled). I need to change the table to some other table (sqlite_sequence) and then back to aaa - only then the columns list is filled.

For me that happens whenever the referenced table is the first in the list. Hope this makes some more sense :smiley: